cli/src/merge_tools/mod.rs

```diff
@@ -172,6 +172,36 @@ fn get_tool_config(
     }
 }
 
+/// Approximate list of known diff tools
+///
+/// The warnings for [[`list_known_diff_editors`]] apply here as well
+pub fn list_known_diff_tools(settings: &UserSettings) -> impl Iterator<Item = &str> {
+    settings.table_keys("merge-tools")
+}
+
+/// Approximate list of known diff editors
+///
+/// Diff tools can be used without configuration. Some merge tools that are
+/// configured for 3-way merging may not work for diffing/diff editing, and we
+/// can't tell which these are. So, this not reliable, but probably good enough
+/// for command-line completion.
+pub fn list_known_diff_editors(settings: &UserSettings) -> impl Iterator<Item = &str> {
+    std::iter::once(":builtin").chain(list_known_diff_tools(settings))
+}
+
+/// List known configured tools for 3-way merging
+pub fn list_known_merge_editors(settings: &UserSettings) -> impl Iterator<Item = &str> {
+    std::iter::once(":builtin").chain(settings.table_keys("merge-tools").filter(|name| {
+        settings
+            .get::<Vec<String>>(ConfigNamePathBuf::from_iter([
+                "merge-tools",
+                name,
+                "merge-args",
+            ]))
+            .is_ok()
+    }))
+}
```

Comment:
Maybe better to inline these in complete.rs? If we add `:ours`/`:theirs`, we might want to attach help text to these.

I think we can fully deserialize `ExternalMergeTool` to test `.merge_args`. The extra cost wouldn't matter.

Reply:
I cleaned it up a bit. I added a new constructor for MergeEditor so that complete.rs does not have to deal with internals, let me know if you have better ideas for naming or the structure. It seemed mildly nicer than having the relevant structures implement Default. It didn't seem worth splitting MergeEditor into two types, but maybe it is?

Reply:
Doesn't `get_external_tool_config()` do the job?

Reply:
`MergeEditor::new` that I'm using does call `get_external_tool_config`, and then also checks that the merge editor is configured.

The naming is confusing.

Reply:
> It seems that `MergeEditor::new` that I'm using does call `get_external_tool_config`, and then also checks that the merge editor is configured. I could call it explicitly and check that the merge-args are non-empty, but I thought the point was to not duplicate that logic (we'll probably need cleverer logic for diff editors and diff tools soon).

I think a bit of duplication is good. We don't need `MergeEditor` abstraction in completion code, and it's weird that we have to create a "dummy".

Reply:
OK, I tried to adjust it to what you asked for (as best I understood it).

I also added a commit to complete diff tools such as `:diff` that were introduced since I made this PR.